### PR TITLE
Document native AOT security aspects

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Cc @dotnet/ilc-contrib 

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/deploying/native-aot/security.md](https://github.com/dotnet/docs/blob/9c551aafee510f7cca9b1f591767564159dfac87/docs/core/deploying/native-aot/security.md) | [docs/core/deploying/native-aot/security](https://review.learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/security?branch=pr-en-us-42585) |

### CODE REVIEW

**Comment by agocke** [comment]
I think it's worth specifically calling out -- no assembly loading. This one seems important because one of the most severe vulnerabilities of the past few years, [Log4Shell](https://en.wikipedia.org/wiki/Log4j#Log4Shell_vulnerability) was specifically due to a Java program being tricked into loading a DLL from a malicious source.

File: docs/core/deploying/native-aot/security.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+author: MichalStrehovsky
+ms.author: michals
+ms.date: 09/11/2024
+---
+
+# Security features
+
+.NET offers many facilities to help address security concerns when building apps. Native AOT deployment builds on top of these facilities and provides several that can help harden your apps.
+
+## No runtime code generation
```
Comment (gewarren) [inline docs/core/deploying/native-aot/security.md:13]:
```suggestion
## No run-time code generation
```

File: docs/core/deploying/native-aot/security.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+## Restricted reflection surface
+
+When apps are published with native AOT, the compiler analyzes the usage of reflection within the app. Only the program elements that were deemed to be targets of reflection are available for reflection at run time. Places within the program that attempt to do unconstrained reflection get flagged using [trimming warnings](../trimming/fixing-warnings.md). Program elements that were not intended to be targets of reflection cannot be reflected on. This can prevent a class of issues where a malicious actor gets in control of what the program reflects on and invokes unintended code. This includes approaches that use `Assembly.LoadFrom` or `Reflection.Emit` - neither of those work with native AOT and their use gets flagged with a warning at build time.
+
+## Control Flow Guard
+
+[Control Flow Guard](/windows/win32/secbp/control-flow-guard) is a highly-optimized platform security feature on Windows that was created to combat memory corruption vulnerabilities. By placing tight restrictions on where an application can execute code from, it makes it much harder for exploits to execute arbitrary code through vulnerabilities such as buffer overflows.
```
Comment (gewarren) [inline docs/core/deploying/native-aot/security.md:23]:
```suggestion
[Control Flow Guard](/windows/win32/secbp/control-flow-guard) is a highly optimized platform security feature on Windows that was created to combat memory corruption vulnerabilities. By placing tight restrictions on where an application can execute code from, it makes it much harder for exploits to execute arbitrary code through vulnerabilities such as buffer overflows.
```

File: docs/core/deploying/native-aot/security.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+To enable Control Flow Guard on you native AOT app set the `ControlFlowGuard` property in the published project.
```
Comment (gewarren) [inline docs/core/deploying/native-aot/security.md:25]:
```suggestion
To enable Control Flow Guard on your native AOT app, set the `ControlFlowGuard` property in the published project.
```

File: docs/core/deploying/native-aot/security.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+## No runtime code generation
+
+Since native AOT generates all code at the time of publishing the app, no new executable code needs to be generated at run time. This allows running your apps in environments that disallow creation of new executable code pages at run time. All the code that the CPU executes can be digitally signed.
+
+## Restricted reflection surface
+
+When apps are published with native AOT, the compiler analyzes the usage of reflection within the app. Only the program elements that were deemed to be targets of reflection are available for reflection at run time. Places within the program that attempt to do unconstrained reflection get flagged using [trimming warnings](../trimming/fixing-warnings.md). Program elements that were not intended to be targets of reflection cannot be reflected on. This can prevent a class of issues where a malicious actor gets in control of what the program reflects on and invokes unintended code. This includes approaches that use `Assembly.LoadFrom` or `Reflection.Emit` - neither of those work with native AOT and their use gets flagged with a warning at build time.
```
Comment (gewarren) [inline docs/core/deploying/native-aot/security.md:19]:
```suggestion
When apps are published with native AOT, the compiler analyzes the usage of reflection within the app. Only the program elements that were deemed to be targets of reflection are available for reflection at run time. Places within the program that attempt to do unconstrained reflection are flagged using [trimming warnings](../trimming/fixing-warnings.md). Program elements that weren't intended to be targets of reflection cannot be reflected on. This restriction can prevent a class of issues where a malicious actor gets in control of what the program reflects on and invokes unintended code. This restriction includes approaches that use `Assembly.LoadFrom` or `Reflection.Emit` - neither of those work with native AOT, and their use is flagged with a warning at build time.
```

**Review by gewarren — APPROVED** [review]

**Pull request merged**